### PR TITLE
Update pulses_avr.cpp

### PR DESCRIPTION
**Pull request by sloped-soarer**
Fix for issue 2938 ... MEGA2560 the first PPM pulse in the frame is longer.
I have tested this, but would like others to test and maybe confirm Bit-bang PPM still works on non Mega2560.

### CODE REVIEW

File: radio/src/pulses/pulses_avr.cpp [hunk 1]
```diff
@@ -110,60 +110,61 @@ uint8_t *pulses2MHzWPtr = pulses2MHz;
...
-  if (IS_DSM2_SERIAL_PROTOCOL(s_current_protocol[0]) || *((uint16_t*)pulses2MHzRPtr) == 0) {
-    if (!IS_DSM2_SERIAL_PROTOCOL(s_current_protocol[0])) {
-      OCR1A = SETUP_PULSES_DURATION;
-    }
-    setupPulses(); // does not sei() for setupPulsesPPM
-    heartbeat |= HEART_TIMER_PULSES;
-    return;
-  }
-
+	uint8_t dt = TCNT1L; // Record Timer1 latency for DEBUG stats display.
```
Comment (projectkk2glider) [inline radio/src/pulses/pulses_avr.cpp]: I will comment on cosmetic errors only. We like to keep code clean and uniform. Use two spaces for indentation.

File: radio/src/pulses/pulses_avr.cpp [hunk 2]
```diff
@@ -110,60 +110,61 @@ uint8_t *pulses2MHzWPtr = pulses2MHz;
...
+
+	// Call setupPulses only after "rest" period has elapsed.
+	// Must do this before toggle PORTB to keep timing accurate.
+	if(IS_DSM2_SERIAL_PROTOCOL(s_current_protocol[0])|| *((uint16_t*)pulses2MHzRPtr) == 0)
+	{
```
Comment (projectkk2glider) [inline radio/src/pulses/pulses_avr.cpp]: Starting brace goes on the line above, except for functions. Like this:
```v

void foo() 
{
  int bar = 6;
  if (bar > 6) {
    bar = 0;
  }
}

```

**Comment by projectkk2glider** [comment]
Thanks for contribution, but please take care of code formatting.

**Comment by bsongis** [comment]
Yes it's really difficult to read, would you rewrite the commit with right indentation? Thanks!

**Comment by sloped-soarer** [comment]
Which code style should i be using ?

**Comment by projectkk2glider** [comment]
Just look at the existing code, most of it should be properly formatted.

**Comment by sloped-soarer** [comment]
Okay ... Thanks. I will try to set up My editor for say K&R with indentation within function definitions.
Indentation of two spaces not tab.
Opening brace on same line ... closing brace indented under related function.
;

**Comment by projectkk2glider** [comment]
You can work in this branch/pull request, just do your edits and push the new commit(s) for inspection.

**Comment by sloped-soarer** [comment]
New line for control statements e.g. 
if(x==2) {
}
else {
}

Any preferences ?

**Comment by projectkk2glider** [comment]
Like that is perfect.

**Comment by sloped-soarer** [comment]
Are You Guys waiting for Me to do something ?.
Apologies for making a mess of the commits. I think i have now configure Eclipse Luna to format the code to the "opentx" style.

File: radio/src/pulses/pulses_avr.cpp [hunk 3]
```diff
@@ -110,60 +110,55 @@ uint8_t *pulses2MHzWPtr = pulses2MHz;
...
-  
-  // Call setupPulses only after REST pulse had been sent.
-  // Must do this before toggle PORTB to keep timing accurate
-  if (IS_DSM2_SERIAL_PROTOCOL(s_current_protocol[0]) || *((uint16_t*)pulses2MHzRPtr) == 0) {
-    if (!IS_DSM2_SERIAL_PROTOCOL(s_current_protocol[0])) {
+  uint8_t dt = TCNT1L; // Record Timer1 latency for DEBUG stats display.
+
+  // Call setupPulses only after "rest" period has elapsed.
+  // Must do this before toggle PORTB to keep timing accurate.
+  if(IS_DSM2_SERIAL_PROTOCOL(s_current_protocol[0])|| *((uint16_t*)pulses2MHzRPtr) == 0) {
```
Comment (projectkk2glider) [inline radio/src/pulses/pulses_avr.cpp]: We prefer space between `if` keyword and parenthesis.

**Comment by projectkk2glider** [comment]
@sloped-soarer we are just busy with other stuff. Please be patient. I have another comment/correction about the style which I missed previously, sorry.

**Comment by tmrttmrt** [comment]
I have executed some short test on a DIY MEGA2560 transmitter.
I used the present pull request rebased on the latest master compiling with: PCB=MEGA2560 LCD=ST7920 AUDIO=ON EXT=FRSKY.
The pulses trace on the oscilloscope appears correct for both positive and negative modulation. A short test with a FrSky receiver also showed no problems. 
I did not review the code since I only vaguely understand how that part of the code works.
Well-done @sloped-soarer .

**Comment by JivaroFAD** [comment]
If all is ok for everyone (right for me), I suggest to merge this pull.

**Comment by projectkk2glider** [comment]
T code changes look good to me, @bsongis  I think you can merge this.

**Comment by sloped-soarer** [comment]
HI @JivaroFAD 
Yes please tell Me how to merge this pull request.

**Comment by projectkk2glider** [comment]
@sloped-soarer you can not merge it, one of the OpenTX developers can do it. We usually let Bertrand do the merges.

**Comment by sloped-soarer** [comment]
@projectkk2glider 
With My history of cock-ups, It's probably better if Bertrand merges it
 :)

File: radio/src/pulses/pulses_avr.cpp [hunk 4]
```diff
@@ -110,58 +110,55 @@ uint8_t *pulses2MHzWPtr = pulses2MHz;
...
+  uint8_t dt = TCNT1L; // Record Timer1 latency for DEBUG stats display.
+
+  // Call setupPulses only after "rest" period has elapsed.
+  // Must do this before toggle PORTB to keep timing accurate.
   if (IS_DSM2_SERIAL_PROTOCOL(s_current_protocol[0]) || *((uint16_t*)pulses2MHzRPtr) == 0) {
     if (!IS_DSM2_SERIAL_PROTOCOL(s_current_protocol[0])) {
       OCR1A = SETUP_PULSES_DURATION;
+#if defined(CPUM2560) // CPUM2560 hardware toggled PPM out.
+      OCR1B = OCR1A;
+      if ( g_model.pulsePol ) TCCR1A = (TCCR1A | (1<<COM1B1)) & ~(1<<COM1B0); // Set idle level.
```
Comment (bsongis) [inline radio/src/pulses/pulses_avr.cpp]: Cosmetics, I would prefer that the test is written differently:
if (xxx) // no spaces
  TCCR1A ...
else
  TCCR1A ...

File: radio/src/pulses/pulses_avr.cpp [hunk 5]
```diff
@@ -110,58 +110,55 @@ uint8_t *pulses2MHzWPtr = pulses2MHz;
...
-      TCCR1A = (3<<COM1B0); // SET the state of PB6(OC1B)/PPM_out on next TCNT1==OCR1B
-      g_ppmPulsePolarity = 0;
-    }
-    else {
-      TCCR1A = (2<<COM1B0); // CLEAR the state of PB6(OC1B)/PPM_out on next TCNT1==OCR1B
-      g_ppmPulsePolarity = 1;
-    }
-  }
+    // CPUM2560 hardware toggled PPM out.
+    if ( *(uint16_t*)(pulses2MHzRPtr + sizeof(uint16_t)) == 0)
```
Comment (bsongis) [inline radio/src/pulses/pulses_avr.cpp]: Same cosmetics here

File: radio/src/pulses/pulses_avr.cpp [hunk 6]
```diff
@@ -110,58 +110,55 @@ uint8_t *pulses2MHzWPtr = pulses2MHz;
...
+    // Look one step ahead to see if we are currently the "rest" period.
+      OCR1B = 0xffff; // Prevent next compare match hence toggle.
+    else OCR1B = *( (uint16_t*) pulses2MHzRPtr);
 #endif
+    }
```
Comment (bsongis) [inline radio/src/pulses/pulses_avr.cpp]: Indentation problem

File: radio/src/pulses/pulses_avr.cpp [hunk 7]
```diff
@@ -110,58 +110,55 @@ uint8_t *pulses2MHzWPtr = pulses2MHz;
...
+
+  OCR1A = *( (uint16_t*) pulses2MHzRPtr); // Schedule next Timer1 interrupt vector (to this function).
+  pulses2MHzRPtr += sizeof(uint16_t); // Non PPM protocols use uint8_t pulse buffer.
 
-  OCR1A = *((uint16_t*)pulses2MHzRPtr); // Schedule next Timer1 interrupt vector (to this function)
```
Comment (bsongis) [inline radio/src/pulses/pulses_avr.cpp]: I prefer this version!

File: radio/src/pulses/pulses_avr.cpp [hunk 8]
```diff
@@ -110,58 +110,55 @@ uint8_t *pulses2MHzWPtr = pulses2MHz;
...
+    // CPUM2560 hardware toggled PPM out.
+    if ( *(uint16_t*)(pulses2MHzRPtr + sizeof(uint16_t)) == 0)
+    // Look one step ahead to see if we are currently the "rest" period.
+      OCR1B = 0xffff; // Prevent next compare match hence toggle.
+    else OCR1B = *( (uint16_t*) pulses2MHzRPtr);
 #endif
+    }
+
+  OCR1A = *( (uint16_t*) pulses2MHzRPtr); // Schedule next Timer1 interrupt vector (to this function).
+  pulses2MHzRPtr += sizeof(uint16_t); // Non PPM protocols use uint8_t pulse buffer.
```
Comment (bsongis) [inline radio/src/pulses/pulses_avr.cpp]: Yes thanks this is consistent with all other comments in this file

**Comment by bsongis** [comment]
It seems all good to me as well, just the little indentation issues.

**Comment by bsongis** [comment]
Thanks for your job!

**Comment by bsongis** [comment]
Would you merge it in 'next' as well? Thanks!

**Comment by projectkk2glider** [comment]
Ported to next in #3536

**Pull request merged**